src/gennylib/include/gennylib/quiesce.hpp

```diff
@@ -0,0 +1,138 @@
+
+#include <string_view>
+
+#include <mongocxx/uri.hpp>
+#include <mongocxx/client.hpp>
+#include <mongocxx/database.hpp>
+#include <bsoncxx/types.hpp>
+#include <bsoncxx/array/view.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+
+#include <boost/log/trivial.hpp>
+
+#include <gennylib/topology.hpp>
+
+namespace genny {
+
+const int DROPPED_COLLECTION_RETRIES = 1000;
+
+bool waitOplog(Topology& topology) {
+    class WaitOplogVisitor : public TopologyVisitor {
+    public:
+        void visitReplSetDescriptionPre(const ReplSetDescription& desc) {
+            using bsoncxx::builder::basic::kvp;
+            using bsoncxx::builder::basic::make_document;
+            mongocxx::client client(mongocxx::uri(desc.primaryUri));
+            auto admin = client["admin"];
+
+            // Assert that all nodes are in reasonable states.
+            auto replSetRes = admin.run_command(make_document(kvp("replSetGetStatus", 1)));
+            auto members = replSetRes.view()["members"];
+            if (members && members.type() == bsoncxx::type::k_array) {
+                bsoncxx::array::view members_view = members.get_array();
+                for (auto member : members_view) {
+                    std::string state(member["stateStr"].get_utf8().value);
+                    if (state != "PRIMARY" && state != "SECONDARY" && state != "ARBITER") {
+                        std::string name(member["name"].get_utf8().value);
+                        BOOST_LOG_TRIVIAL(error) << "Cannot wait oplog, replset member "
+                            << name << " is " << state;
+                        success_acc = false;
+                        return;
+                    }
+                }
+            }
+
+            // Do flush
+            auto collection = admin["wait_oplog"];
+            mongocxx::write_concern wc;
+            wc.nodes(desc.nodes.size());
+            wc.journal(true);
+            mongocxx::options::insert opts;
+            opts.write_concern(wc);
+            auto insertRes = collection.insert_one(make_document(kvp("x", "flush")), opts);
+        
+            success_acc = success_acc && insertRes && insertRes->result().inserted_count() == 1;
+        }
+        bool success_acc = true;
+    };
+
+    WaitOplogVisitor waitVisitor;
+    topology.accept(waitVisitor);
+
+    return waitVisitor.success_acc;
+}
+
+void doFsync(Topology& topology) {
+    class DoFsyncVisitor : public TopologyVisitor {
+        void visitMongodDescription(const MongodDescription& desc) {
+            using bsoncxx::builder::basic::kvp;
+            using bsoncxx::builder::basic::make_document;
+
+            mongocxx::client client(mongocxx::uri(desc.mongodUri));
+            auto admin = client["admin"];
+            admin.run_command(make_document(kvp("fsync", 1)));
+        }
+    };
+    DoFsyncVisitor v;
+    topology.accept(v);
+}
+
+bool checkForDroppedCollections(mongocxx::database db) {
```

Comment:
This logic is different for 4.0 and earlier, but my understanding is that we use legacy genny on the legacy DSI branches, so I assumed adding a version check here would be unnecessary?

Reply:
The `legacy` branch is going away so all branches will be on genny master. It's not unreasonable to say that workloads that use this can't be run on 4.0. If anyone balks we can perhaps do runtime version checks.